test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	crdv1alpha2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	. "antrea.io/antrea/test/e2e/utils"
+)
+
+type VmInfo struct {
+	Nodename string
+	OS       string
+	IfName   string
+	IP       string
+	IfIndex  string // Used only for windows
+}
+
+// TestBasic is the top-level test which contains some subtests for
+// basic test cases so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfHasWindowsNodes(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	t.Run("testExternalEntityCommon", func(t *testing.T) { testExternalEntityCommon(t, data, "test-ns") })
+}
+
+func testExternalEntityCommon(t *testing.T, data *TestData, namespace string) {
+	t.Logf("List of Windows VMs: '%s'", testOptions.winVMs)
+	t.Logf("List of Linux VMs: '%s'", testOptions.linVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), namespace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(namespace).Get(context.TODO(), "vm-agent", metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using name space %s", nsReturned.Name)
+	var VmList []VmInfo
+	if testOptions.linVMs != "" {
+		vms := strings.Split(testOptions.linVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for linux VM: %s", vm)
+			vmInfo := GetVmInfo(t, data, vm)
+			VmList = append(VmList, vmInfo)
+		}
+	}
+
+	if testOptions.winVMs != "" {
+		vms := strings.Split(testOptions.winVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for windows VM, %s", vm)
+			vmInfo := GetWindowsVmInfo(t, data, vm)
+			VmList = append(VmList, vmInfo)
+		}
+	}
+
+	t.Log("=== TestSetup ===")
+	for _, vm := range VmList {
+		// Stop antrea-agent service
+		StopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalEntity for %+v", vm)
+		_, err := CreateExternalEntityCrd(data, nsReturned.Name, vm.Nodename, vm.IfName, vm.IP)
```

Comment:
No need to create ExternalEntity since the Controller logic to convert ExternalNode to ExternalEntity is already added? @mengdie-song What is your point?

Reply:
Ok. I will update it.